# Unit 12 - Tales from the Crypto

---


## 1. Sentiment Analysis

Use the [newsapi](https://newsapi.org/) to pull the latest news articles for Bitcoin and Ethereum and create a DataFrame of sentiment scores for each coin.

Use descriptive statistics to answer the following questions:
1. Which coin had the highest mean positive score?
2. Which coin had the highest negative score?
3. Which coin had the highest positive score?

In [21]:
# Initial imports
import os
import pandas as pd
from dotenv import load_dotenv
import nltk as nltk
nltk.download('vader_lexicon')
from nltk.sentiment.vader import SentimentIntensityAnalyzer
analyzer = SentimentIntensityAnalyzer()

%matplotlib inline

from newsapi import NewsApiClient
load_dotenv()

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     /Users/emilioacubero/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


True

In [22]:
# Read your api key environment variable
api_key = os.getenv("news_api")

In [23]:
# Create a newsapi client
newsapi = NewsApiClient(api_key=api_key)

In [24]:
# Fetch the Bitcoin news articles
bitcoin_news = newsapi.get_everything(q='Bitcoin', language='en')

In [25]:
# Fetch the Ethereum news articles
ethereum_news = newsapi.get_everything(q='Ethereum', language='en')

In [26]:
# Create the Bitcoin sentiment scores DataFrame
bitcoin_sentiments = []

for article in bitcoin_news["articles"]:
    try:
        text = article["content"]
        date = article["publishedAt"][:10]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]
        pos = sentiment["pos"]
        neu = sentiment["neu"]
        neg = sentiment["neg"]
        
        bitcoin_sentiments.append({
            "text": text,
            "date": date,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu
            
        })
        
    except AttributeError:
        pass
    
bitcoin_df = pd.DataFrame(bitcoin_sentiments)

cols = ["date", "text", "compound", "positive", "negative", "neutral"]
bitcoin_df = bitcoin_df[cols]

bitcoin_df.head()

date                                               text  compound  \
0  2022-02-17  Even in cyberspace, the Department of Justice ...    0.7351   
1  2022-02-13  The couple would never flee from the country a...   -0.3400   
2  2022-02-14  If Sundays Super Bowl ads are any indication, ...    0.7783   
3  2022-02-15  The confiscation of the Bitcoin that remained ...    0.6486   
4  2022-03-12  "Bitcoin was seen by many of its libertarian-l...   -0.7713   

   positive  negative  neutral  
0     0.147     0.000    0.853  
1     0.057     0.118    0.825  
2     0.163     0.000    0.837  
3     0.146     0.000    0.854  
4     0.000     0.169    0.831

In [27]:
# Create the Ethereum sentiment scores DataFrame
ethereum_sentiments = []

for article in ethereum_news["articles"]:
    try:
        text = article["content"]
        date = article["publishedAt"][:10]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]
        pos = sentiment["pos"]
        neu = sentiment["neu"]
        neg = sentiment["neg"]
        
        ethereum_sentiments.append({
            "text": text,
            "date": date,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu
            
        })
        
    except AttributeError:
        pass
    
# Create DataFrame
ethereum_df = pd.DataFrame(ethereum_sentiments)

# Reorder DataFrame columns
cols = ["date", "text", "compound", "positive", "negative", "neutral"]
ethereum_df = ethereum_df[cols]

ethereum_df.head()

date                                               text  compound  \
0  2022-03-01  In February, shit hit the fan in the usual way...   -0.3182   
1  2022-02-17  Technical analysis isnt a perfect tool, but it...   -0.2498   
2  2022-02-16  People in Colorado will be able to pay their s...   -0.1027   
3  2022-02-25  Coinbase reported that the share of trading vo...    0.6705   
4  2022-02-26  Illustration by James Bareham / The Verge\r\n\...   -0.4588   

   positive  negative  neutral  
0     0.059     0.093    0.848  
1     0.000     0.059    0.941  
2     0.000     0.036    0.964  
3     0.188     0.000    0.812  
4     0.000     0.083    0.917

In [28]:
# Describe the Bitcoin Sentiment
bitcoin_df.describe()

compound   positive   negative    neutral
count  20.000000  20.000000  20.000000  20.000000
mean    0.064760   0.069600   0.049900   0.880500
std     0.476352   0.051707   0.065011   0.066818
min    -0.771300   0.000000   0.000000   0.739000
25%    -0.277550   0.044000   0.000000   0.835500
50%     0.038600   0.066000   0.000000   0.870500
75%     0.455475   0.094250   0.097750   0.935750
max     0.778300   0.163000   0.187000   1.000000

In [29]:
# Describe the Ethereum Sentiment
ethereum_df.describe()

compound   positive   negative    neutral
count  20.000000  20.000000  20.000000  20.000000
mean    0.155640   0.076150   0.032200   0.891600
std     0.414052   0.083462   0.040875   0.090565
min    -0.526700   0.000000   0.000000   0.692000
25%    -0.110175   0.000000   0.000000   0.841000
50%     0.000000   0.057500   0.000000   0.904500
75%     0.528675   0.130250   0.059000   0.958000
max     0.834000   0.249000   0.115000   1.000000

### Questions:

Q: Which coin had the highest mean positive score?

A: Ethereum has the highest mean positive score with 0.076.
 
Q: Which coin had the highest compound score?

A: Ethereum has the highest compund score with 0.834.

Q. Which coin had the highest positive score?

A: Ethereum has the highest positive score with 0.249.

---

## 2. Natural Language Processing
---
###   Tokenizer

In this section, you will use NLTK and Python to tokenize the text for each coin. Be sure to:
1. Lowercase each word.
2. Remove Punctuation.
3. Remove Stopwords.

In [30]:
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer, PorterStemmer
from string import punctuation
import re

In [35]:
# Instantiate the lemmatizer
lemmatizer = WordNetLemmatizer()

# Create a list of stopwords
sw = set(stopwords.words('english'))

# Expand the default stopwords list if necessary
sw_addon = {'said', 'sent', 'found', 'including', 'today', 'announced', 'week', 'basically', 'ha', 'char','also', 'reutersdado'}

In [36]:
# Complete the tokenizer function

def tokenizer(article):
    sw = set(stopwords.words('english'))
    regex = re.compile("[^a-zA-Z ]")
    re_clean = regex.sub('', article)
    words = word_tokenize(re_clean)
    lem = [lemmatizer.lemmatize(word) for word in words]
    output = [word.lower() for word in lem if word.lower() not in sw.union(sw_addons)]
    return output
    
    
    return tokens

In [38]:
# Create a new tokens column for Bitcoin
btc_word_tokenized = []
for text in bitcoin_df.text:  
    words = []
    words = words + tokenizer(text)
    btc_word_tokenized.append(words)

bitcoin_df["tokens"]=btc_word_tokenized
bitcoin_df.head()

date                                               text  compound  \
0  2022-02-17  Even in cyberspace, the Department of Justice ...    0.7351   
1  2022-02-13  The couple would never flee from the country a...   -0.3400   
2  2022-02-14  If Sundays Super Bowl ads are any indication, ...    0.7783   
3  2022-02-15  The confiscation of the Bitcoin that remained ...    0.6486   
4  2022-03-12  "Bitcoin was seen by many of its libertarian-l...   -0.7713   

   positive  negative  neutral  \
0     0.147     0.000    0.853   
1     0.057     0.118    0.825   
2     0.163     0.000    0.837   
3     0.146     0.000    0.854   
4     0.000     0.169    0.831   

                                              tokens  
0  [even, cyberspace, department, justice, able, ...  
1  [couple, would, never, flee, country, risk, lo...  
2  [sundays, super, bowl, ad, indication, crypto,...  
3  [confiscation, bitcoin, remained, mr, lichtens...  
4  [bitcoin, wa, seen, many, libertarianleaning, ...

In [39]:
# Create a new tokens column for Ethereum
eth_word_tokenized = []
for text in ethereum_df.text:
    words = []
    words = words + tokenizer(text)
    eth_word_tokenized.append(words)
    
ethereum_df["tokens"]=eth_word_tokenized
ethereum_df.head()

date                                               text  compound  \
0  2022-03-01  In February, shit hit the fan in the usual way...   -0.3182   
1  2022-02-17  Technical analysis isnt a perfect tool, but it...   -0.2498   
2  2022-02-16  People in Colorado will be able to pay their s...   -0.1027   
3  2022-02-25  Coinbase reported that the share of trading vo...    0.6705   
4  2022-02-26  Illustration by James Bareham / The Verge\r\n\...   -0.4588   

   positive  negative  neutral  \
0     0.059     0.093    0.848   
1     0.000     0.059    0.941   
2     0.000     0.036    0.964   
3     0.188     0.000    0.812   
4     0.000     0.083    0.917   

                                              tokens  
0  [february, shit, hit, fan, usual, way, old, tw...  
1  [technical, analysis, isnt, perfect, tool, may...  
2  [people, colorado, able, pay, state, tax, cryp...  
3  [coinbase, reported, share, trading, volume, e...  
4  [illustration, james, bareham, verge, million,...

---

### NGrams and Frequency Analysis

In this section you will look at the ngrams and word frequency for each coin. 

1. Use NLTK to produce the n-grams for N = 2. 
2. List the top 10 words for each coin. 

In [15]:
from collections import Counter
from nltk import ngrams

In [16]:
# Generate the Bitcoin N-grams where N=2
# YOUR CODE HERE!

In [17]:
# Generate the Ethereum N-grams where N=2
# YOUR CODE HERE!

In [18]:
# Function token_count generates the top 10 words for a given coin
def token_count(tokens, N=3):
    """Returns the top N tokens from the frequency count"""
    return Counter(tokens).most_common(N)

In [19]:
# Use token_count to get the top 10 words for Bitcoin
# YOUR CODE HERE!

In [20]:
# Use token_count to get the top 10 words for Ethereum
# YOUR CODE HERE!

---

### Word Clouds

In this section, you will generate word clouds for each coin to summarize the news for each coin

In [21]:
from wordcloud import WordCloud
import matplotlib.pyplot as plt
plt.style.use('seaborn-whitegrid')
import matplotlib as mpl
mpl.rcParams['figure.figsize'] = [20.0, 10.0]

In [22]:
# Generate the Bitcoin word cloud
# YOUR CODE HERE!

In [23]:
# Generate the Ethereum word cloud
# YOUR CODE HERE!

---
## 3. Named Entity Recognition

In this section, you will build a named entity recognition model for both Bitcoin and Ethereum, then visualize the tags using SpaCy.

In [24]:
import spacy
from spacy import displacy

In [25]:
# Download the language model for SpaCy
# !python -m spacy download en_core_web_sm

In [26]:
# Load the spaCy model
nlp = spacy.load('en_core_web_sm')

---
### Bitcoin NER

In [27]:
# Concatenate all of the Bitcoin text together
# YOUR CODE HERE!

In [28]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [29]:
# Render the visualization
# YOUR CODE HERE!

In [30]:
# List all Entities
# YOUR CODE HERE!

---

### Ethereum NER

In [31]:
# Concatenate all of the Ethereum text together
# YOUR CODE HERE!

In [32]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [33]:
# Render the visualization
# YOUR CODE HERE!

In [34]:
# List all Entities
# YOUR CODE HERE!

---